core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -475,7 +475,7 @@ private[log] class Cleaner(val id: Int,
       if (writeBuffer.position > 0) {
         writeBuffer.flip()
         val retained = MemoryRecords.readableRecords(writeBuffer)
-        dest.append(firstOffset = retained.deepEntries.iterator.next().offset,
+        dest.append(firstOffset = retained.batches.iterator.next().baseOffset,
```

Comment:
See MemoryRecordsTest.filterTo. The test is kind of a bear, but it covers a large number of cases. I'll update the doc for `lastOffset` and `maxTimestamp`.